File: gov_executive_platform/functions/test/daily_update_delete.test.mjs
// من يحذف تحديثاً يومياً — على الخادم.
//
// الحذف يقع بصلاحية المدير (لأن معه محوَ ملفاتٍ من التخزين لا تأذن به
// قواعدُ التخزين)، فيتجاوز قواعد Firestore. فما يُقاس هنا هو أن النصّ
// المُعاد على الخادم **هو نفسه** نصُّ القاعدة، على الحالات نفسها.
//
// ونظيرُه على المحاكي في `test_rules/daily_update.rules.test.mjs`،
// ونظيرُه في العميل في `test/daily_update_delete_test.dart`. ثلاثةُ
// مواضع لنصٍّ واحد — فإن افترقت، افترق ما يراه المستخدم عمّا يقع فعلاً.
import {test, describe} from "node:test";
import assert from "node:assert/strict";
import {mayDeleteDailyUpdate} from "../lib/daily_update_delete.js";

const DEPT = "d-1";
const project = (extra = {}) => ({
  departmentId: DEPT,
  managerUids: ["m1", "m2"],
  executorUids: ["e1"],
  ...extra,
});
const update = (authorUid = "m1") => ({authorUid});

const actor = (uid, role, departmentIds = [DEPT]) => ({uid, role, departmentIds});

describe("من يحذف", () => {
  test("مسؤول النظام — أيّ تحديث", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("a", "systemAdmin", []), update(), project()),
      true,
    );
  });

  test("وكاتبُه — ولو كان موظفاً منفّذاً", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("e1", "employee"), update("e1"), project()),
      true,
    );
  });

  test("ومديرُ المشروع الأول على تحديثٍ لم يكتبه", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("m1", "projectOfficer"), update("e1"), project()),
      true,
    );
  });

  // نظير عطل الاشتراك والدمج: القائمة لا المفرد.
  test("والمدير الثاني كذلك — بالقائمة لا بالمفرد", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("m2", "projectOfficer"), update("m1"), project()),
      true,
    );
  });

  test("ومديرُ الإدارة صاحبتِه", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("h", "departmentManager"), update(), project()),
      true,
    );
  });

  test("ومديرُ إدارةٍ يديرُ أكثر من واحدة، منها إدارةُ المشروع", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("h", "departmentManager", ["d-9", DEPT]), update(), project()),
      true,
    );
  });
});

describe("ومن لا يحذف", () => {
  // حدُّ الفتح: المنفّذ **يكتب** التحديث اليومي ولا يمحو تحديث مديره.
  test("المنفّذُ تحديثَ مديره", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("e1", "employee"), update("m1"), project()),
      false,
    );
  });

  test("وزميلٌ في الإدارة ليس عضواً", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("z", "employee"), update("m1"), project()),
      false,
    );
  });

  test("ومديرُ إدارةٍ أخرى", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("h", "departmentManager", ["d-9"]), update(), project()),
      false,
    );
  });

  // يرى كل الإدارات ولا يغيّر فيها شيئاً.
  test("والمستخدم التنفيذي", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("v", "executiveViewer", []), update(), project()),
      false,
    );
  });

  // ــ والحالة التي بها وحدها يُقاس هذا الشرط ــ
  //
  // المستخدم التنفيذي لا يمرّ بالفروع الأخرى أصلاً (ليس كاتباً ولا مديراً
  // ولا مدير إدارة)، فإسقاطُ شرطه لا يغيّر النتيجة — **إلا إن كان هو
  // الكاتب**. وذلك يقع فعلاً: موظفٌ كتب تحديثاته ثم نُقل إلى هذا الدور،
  // فبقي اسمُه عليها. وبعد النقل هو ناظرٌ لا عامل، فلا يمحو.
  test("ولو كان هو كاتبَ التحديث — نُقل إلى الدور بعد أن كتب", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("v", "executiveViewer", []), update("v"), project()),
      false,
    );
  });

  test("وبلا معرّف حساب", () => {
    assert.equal(mayDeleteDailyUpdate(actor("", "systemAdmin", []), update(), project()), false);
  });
});

describe("ولا ينكسر على مستندٍ ناقص", () => {
  // مشروعٌ حُذف بينما نافذةُ اليوم مفتوحة: يبقى للكاتب حقُّه ولا يُخمَّن غيره.
  test("مشروعٌ غير موجود — للكاتب وحده", () => {
    assert.equal(mayDeleteDailyUpdate(actor("e1", "employee"), update("e1"), null), true);
    assert.equal(mayDeleteDailyUpdate(actor("m1", "projectOfficer"), update("e1"), null), false);
  });

  test("ومشروعٌ قديم بلا قائمة مديرين", () => {
    assert.equal(
      mayDeleteDailyUpdate(actor("m1", "projectOfficer"), update("e1"), {departmentId: DEPT}),
      false,
    );
  });

  test("ومديرُ إدارةٍ بلا إدارات في بصمته", () => {
    assert.equal(
      mayDeleteDailyUpdate({uid: "h", role: "departmentManager"}, update(), project()),
      false,
    );
  });
});

File: gov_executive_platform/test/daily_update_delete_test.dart
// نافذة اليوم: من يرى فيها ماذا، ومن يملك زرّ الحذف.
//
// ــــ عطلان يقيسهما هذا الملف ــــ
//
// **الأول**: الضغط على يومٍ في التقويم كان يقذف من يملك التحرير — أي مدير
// المشروع نفسه — إلى نموذج إضافةٍ فارغ، ولا يريه ما كُتب في ذلك اليوم
// إطلاقاً. ومن لا يملك التحرير كان يرى السرد. فصاحبُ المشروع وحده هو
// المحروم من سجلّه. وهذا نصُّ ما اشتُكي منه: «يجب أن يظهر له التحديث الذي
// قام بإضافته ويكون قادراً على عرضه وحذفه».
//
// **والثاني**: لم يكن للتحديث حذفٌ إطلاقاً — لا دالّة ولا زرّ، والقاعدة
// `allow delete: if isAdmin()`. فمن كتب تحديثاً بالخطأ لا سبيل له إلى محوه.
//
// وحدُّ الحذف هو ما يُقاس أكثر من الحذف نفسه: **المنفّذ لا يمحو تحديث
// مديره**، ولا يمحوه زميلٌ في الإدارة. راجع نظيره على المحاكي في
// `test_rules/daily_update.rules.test.mjs` — القاعدة هي الحَكَم، وهذه
// الواجهة مرآتُها.
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:provider/provider.dart';

import 'package:gov_exec_platform/data/app_store.dart';
import 'package:gov_exec_platform/models/app_user.dart';
import 'package:gov_exec_platform/models/attachment.dart';
import 'package:gov_exec_platform/models/daily_update.dart';
import 'package:gov_exec_platform/models/enums.dart';
import 'package:gov_exec_platform/models/project.dart';
import 'package:gov_exec_platform/theme/app_theme.dart';
import 'package:gov_exec_platform/widgets/day_updates_dialog.dart';

const _dept = 'd-1';
final _day = DateTime(2026, 6, 15);

AppUser _user(String id, UserRole role, {String? dept = _dept}) => AppUser(
      id: id,
      name: 'مستخدم $id',
      email: '$id@moj.gov.kw',
      phone: '',
      role: role,
      departmentId: dept,
      departmentIds: dept == null ? const [] : [dept],
      status: UserStatus.approved,
      createdAt: DateTime(2026, 1, 1),
    );

Project _project() => Project(
      id: 'p1',
      departmentId: _dept,
      name: 'مشروع العدالة الرقمية',
      description: '',
      startDate: DateTime(2026, 1, 1),
      dueDate: DateTime(2026, 12, 31),
      status: ProjectStatus.onTrack,
      priority: PriorityLevel.medium,
      progressPercent: 20,
      managerUids: const ['m1', 'm2'],
      executorUids: const ['e1'],
    );

DailyUpdate _update({
  String id = 'u1',
  String authorUid = 'm1',
  String authorName = 'المدير الأول',
  List<Attachment> attachments = const [],
}) =>
    DailyUpdate(
      id: id,
      projectId: 'p1',
      departmentId: _dept,
      authorUid: authorUid,
      authorName: authorName,
      date: _day,
      achievements: 'أُنجز ربط النظام بالمحاكم',
      completedTasks: const [],
      newRisks: const [],
      blockers: const [],
      decisionsRequired: const [],
      progressPercent: 40,
      attachments: attachments,
    );

AppStore _store(AppUser me, {List<DailyUpdate>? updates}) => AppStore()
  ..currentUser = me
  ..projects = [_project()]
  ..dailyUpdates = updates ?? [_update()];

Future<void> _pump(WidgetTester tester, AppStore store) async {
  await tester.pumpWidget(MaterialApp(
    theme: AppTheme.theme,
    home: ChangeNotifierProvider<AppStore>.value(
      value: store,
      child: Directionality(
        textDirection: TextDirection.rtl,
        child: Scaffold(
          body: DayUpdatesDialog(project: _project(), day: _day),
        ),
      ),
    ),
  ));
  await tester.pump();
}

void main() {
  group('نافذة اليوم تعرض ما جرى فيه', () {
    // العطل الأول بعينه: من يملك التحرير كان لا يرى شيئاً.
    testWidgets('لمدير المشروع — لا نموذج إضافةٍ فارغ', (tester) async {
      await _pump(tester, _store(_user('m1', UserRole.projectOfficer)));
      expect(tester.takeException(), isNull);

      expect(find.text('أُنجز ربط النظام بالمحاكم'), findsOneWidget);
      expect(find.text('المدير الأول'), findsOneWidget);
      // وزرُّ الإضافة موجودٌ مع ذلك — عرضٌ **ثم** إضافة، لا إحداهما بدل الأخرى.
      expect(find.text('أضِف تحديثاً لهذا اليوم'), findsOneWidget);
    });

    testWidgets('وللمدير الثاني كذلك', (tester) async {
      await _pump(tester, _store(_user('m2', UserRole.projectOfficer)));
      expect(find.text('أُنجز ربط النظام بالمحاكم'), findsOneWidget);
    });

    testWidgets('ومن لا يكتب يقرأ بلا زرّ إضافة', (tester) async {
      await _pump(tester, _store(_user('x', UserRole.executiveViewer, dept: null)));
      expect(find.text('أُنجز ربط النظام بالمحاكم'), findsOneWidget);
      expect(find.text('أضِف تحديثاً لهذا اليوم'), findsNothing);
    });

    testWidgets('ويومٌ فارغ يُقال إنه فارغ', (tester) async {
      await _pump(tester, _store(_user('m1', UserRole.projectOfficer), updates: []));
      expect(find.text('لا يوجد تحديث مسجَّل لهذا اليوم.'), findsOneWidget);
    });

    testWidgets('والمرفقات تُعرض ليُفتح منها', (tester) async {
      final store = _store(
        _user('m1', UserRole.projectOfficer),
        updates: [
          _update(attachments: const [
            Attachment(name: 'محضر.pdf', url: 'https://x/1', kind: AttachmentKind.upload),
          ])
        ],
      );
      await _pump(tester, store);
      expect(find.text('محضر.pdf'), findsOneWidget);
    });
  });

  group('وزرّ الحذف لمن يملكه', () {
    testWidgets('كاتبُ التحديث يراه', (tester) async {
      await _pump(
        tester,
        _store(_user('e1', UserRole.employee), updates: [_update(authorUid: 'e1')]),
      );
      expect(find.byTooltip('حذف هذا التحديث'), findsOneWidget);
    });

    testWidgets('ومديرُ المشروع يراه على تحديثٍ لم يكتبه', (tester) async {
      await _pump(tester, _store(_user('m2', UserRole.projectOfficer)));
      expect(find.byTooltip('حذف هذا التحديث'), findsOneWidget);
    });

    testWidgets('ومديرُ الإدارة كذلك', (tester) async {
      await _pump(tester, _store(_user('h', UserRole.departmentManager)));
      expect(find.byTooltip('حذف هذا التحديث'), findsOneWidget);
    });

    // حدُّ الفتح: المنفّذ **يكتب** التحديث اليومي ولا يمحو تحديث مديره.
    testWidgets('ولا يراه المنفّذ على تحديث مديره', (tester) async {
      await _pump(tester, _store(_user('e1', UserRole.employee)));
      expect(find.text('أُنجز ربط النظام بالمحاكم'), findsOneWidget);
      expect(find.byTooltip('حذف هذا التحديث'), findsNothing);
    });

    testWidgets('ولا زميلٌ في الإدارة ليس عضواً', (tester) async {
      await _pump(tester, _store(_user('z', UserRole.employee)));
      expect(find.byTooltip('حذف هذا التحديث'), findsNothing);
    });

    // المستخدم التنفيذي يرى كل شيء ولا يغيّر شيئاً — قاعدةٌ قائمة في المنصة.
    testWidgets('ولا المستخدم التنفيذي وإن رأى كل الإدارات', (tester) async {
      await _pump(tester, _store(_user('v', UserRole.executiveViewer, dept: null)));
      expect(find.byTooltip('حذف هذا التحديث'), findsNothing);
    });
  });

  // نفس القرار بلا واجهة — فهو ما تقرؤه الشاشة، وما يُقارَن بالقاعدة.
  group('canDeleteDailyUpdate', () {
    test('مسؤول النظام يحذف أيّ تحديث', () {
      final store = _store(_user('a', UserRole.systemAdmin, dept: null));
      expect(store.canDeleteDailyUpdate(_update(), _project()), isTrue);
    });

    test('وبلا مشروعٍ محمَّل يبقى للكاتب حقُّه', () {
      final store = _store(_user('e1', UserRole.employee));
      expect(store.canDeleteDailyUpdate(_update(authorUid: 'e1'), null), isTrue);
      expect(store.canDeleteDailyUpdate(_update(authorUid: 'm1'), null), isFalse);
    });

    test('ومديرُ إدارةٍ أخرى لا يحذف', () {
      final store = _store(_user('h', UserRole.departmentManager, dept: 'd-9'));
      expect(store.canDeleteDailyUpdate(_update(), _project()), isFalse);
    });
  });
}
